### PR TITLE
Print module

### PR DESCRIPTION
**Pull request by TurboGit**
This is a first version for review. I propose it a bit earlier than before as the gtk3 is now merged it is harder for me to keep this branch up-to-date with master.

As proposed by houz I'll be happy to get help to migrate this piece of code to Gtk3.

I'll be also happy to get comment from review to straighten the code where needed.

Thanks.

### CODE REVIEW

File: src/common/cups_print.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+*/
+
+#ifndef DT_CUPS_PRINT_H
+#define DT_CUPS_PRINT_H
+
+#include <iop/color.h>
+
+#define MAX_NAME 128
+
+typedef enum _alignement
```
Comment (hanatos) [inline src/common/cups_print.h]: not that it matters.. but this looks like a typo (french spelling?)

File: src/common/printprof.c [hunk 2]
```diff
@@ -0,0 +1,763 @@
...
+
+#include <glib.h>
+#include <unistd.h>
+#include "lcms2_plugin.h"
+#include "tiffio.h"
+#include "common/printprof.h"
+
+#define SW "-"
+
+typedef struct _print_profile_data_t
```
Comment (hanatos) [inline src/common/printprof.c]: i think you can actually give the struct the same name and the same tag. why the leading underscore?

File: src/control/jobs/control_jobs.c [hunk 3]
```diff
@@ -1033,6 +1034,50 @@ static int32_t dt_control_export_job_run(dt_job_t *job)
...
+  dt_control_export_job_run(job);
+
+  // now apply the printer profile/intent if any
+
+  if (strlen(pp->pinfo.printer.profile)>0 && g_file_test(pp->pinfo.printer.profile,G_FILE_TEST_EXISTS))
+  {
+    fprintf(stderr, "[print] printer profile %s and intent %d\n", pp->pinfo.printer.profile, pp->pinfo.printer.intent);
+    if (!dt_apply_printer_profile(filename, pp->pinfo.printer.profile, pp->pinfo.printer.intent))
+    {
+      dt_control_log(_("failed to apply printer profile '%'"), pp->pinfo.printer.profile);
```
Comment (hanatos) [inline src/control/jobs/control_jobs.c]: missing format character after % ?

File: src/common/cups_print.h (hunk 1 above)
Comment (TurboGit) [inline src/common/cups_print.h]: indeed, a french spelling! fixed.

File: src/common/printprof.c (hunk 2 above)
Comment (TurboGit) [inline src/common/printprof.c]: right, fixed here and in other places.

File: src/control/jobs/control_jobs.c (hunk 3 above)
Comment (TurboGit) [inline src/control/jobs/control_jobs.c]: good catch, %s of course here. fixed.

File: CMakeLists.txt [hunk 4]
```diff
@@ -30,6 +30,7 @@ option(USE_OPENJPEG "Enable JPEG 2000 support" ON)
 option(USE_WEBP "Enable WebP export support" ON)
 option(BUILD_CMSTEST "Build a test program to check your system's color management setup" ON)
 option(USE_OPENEXR "Enable OpenEXR support" ON)
+option(USE_PRINT "Build the print module" ON)
```
Comment (LebedevRI) [inline CMakeLists.txt]: I think it should be BUILD_PRINT, just like BUILD_SLIDESHOW

File: src/libs/print_settings.c [hunk 5]
```diff
@@ -0,0 +1,1168 @@
...
+  d->b_left   = gtk_spin_button_new_with_range(0, 10000, 1);
+  d->b_right  = gtk_spin_button_new_with_range(0, 10000, 1);
+  d->b_bottom = gtk_spin_button_new_with_range(0, 10000, 1);
+
+  ////////////////////////// PRINTER SETTINGS
+
+  // create papers combo as filled when adding printers
+  d->papers = GTK_COMBO_BOX(gtk_combo_box_text_new());
+
+  GList *printers = dt_get_printers();
```
Comment (LebedevRI) [inline src/libs/print_settings.c]: I do not see where this is being freed (what was allocated in dt_get_printer_info())

File: src/common/cups_print.h [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+#define MAX_NAME 128
+
+typedef enum dt_alignment_t
+{
+  top_left, top, top_right,
+  left, center, right,
+  bottom_left, bottom, bottom_right
+} dt_alignment_t;
+
+typedef struct dt__paper_info_t
```
Comment (LebedevRI) [inline src/common/cups_print.h]: You do not need __ there, dt_paper_info_t should work just fine.

File: src/common/printprof.c [hunk 7]
```diff
@@ -0,0 +1,761 @@
...
+  int PrecalcMode ;
+
+  cmsFloat64Number ObserverAdaptationState;
+
+  const char *cInpProf;
+  const char *cOutProf;
+  TIFF *in, *out;
+} dt_print_profile_data_t;
+
+#define FatalError printf
```
Comment (LebedevRI) [inline src/common/printprof.c]: Why not use printf() directly?
Alternatively, maybe you should call dt_control_log()

File: src/control/jobs/control_jobs.c [hunk 8]
```diff
@@ -1033,6 +1034,50 @@ static int32_t dt_control_export_job_run(dt_job_t *job)
...
 }
 
+#ifdef HAVE_PRINT
+static int32_t dt_control_print_job_run(dt_job_t *job)
+{
+  dt_control_image_enumerator_t *params = (dt_control_image_enumerator_t *)dt_control_job_get_params(job);
+  GList *t = params->index;
+  dt_control_print_t *pp = (dt_control_print_t*)params->data;
+  int32_t imgid = 0;
+  char filename[PATH_MAX];
```
Comment (LebedevRI) [inline src/control/jobs/control_jobs.c]: = { 0 }

File: src/control/jobs/control_jobs.c [hunk 9]
```diff
@@ -1033,6 +1034,50 @@ static int32_t dt_control_export_job_run(dt_job_t *job)
...
+
+  if (t) imgid = GPOINTER_TO_INT(t->data);
+
+  g_strlcpy(filename, pp->filename, PATH_MAX);
```
Comment (LebedevRI) [inline src/control/jobs/control_jobs.c]: sizeof(pp->filename)

File: src/control/jobs/control_jobs.c [hunk 10]
```diff
@@ -1033,6 +1034,50 @@ static int32_t dt_control_export_job_run(dt_job_t *job)
...
+  g_strlcat(filename, ".tif", PATH_MAX);
```
Comment (LebedevRI) [inline src/control/jobs/control_jobs.c]: sizeof(filename)

File: src/control/jobs/control_jobs.c [hunk 11]
```diff
@@ -1033,6 +1034,50 @@ static int32_t dt_control_export_job_run(dt_job_t *job)
...
+
+  // and finaly print
+
+  dt_print_file (imgid, filename, &pp->pinfo);
+
+  // add tag for this image
+
+  char tag[256] = { 0 };
+  guint tagid = 0;
+  snprintf (tag, 256, "darktable|printed|%s", pp->pinfo.printer.name);
```
Comment (LebedevRI) [inline src/control/jobs/control_jobs.c]: sizeof(tag)

File: src/control/jobs/control_jobs.c [hunk 12]
```diff
@@ -1319,6 +1364,56 @@ void dt_control_export(GList *imgid_list, int max_width, int max_height, int for
...
+    return;
+  }
+  dt_control_job_set_params(job, params);
+  params->index = imgid_list;
+  dt_control_print_t *data = (dt_control_print_t*)malloc(sizeof(dt_control_print_t));
+  data->max_width = max_width;
+  data->max_height = max_height;
+  data->format_index = format_index;
+  data->storage_index = storage_index;
+  g_strlcpy(data->filename, filename, PATH_MAX);
```
Comment (LebedevRI) [inline src/control/jobs/control_jobs.c]: sizeof()

File: src/control/jobs/control_jobs.c [hunk 13]
```diff
@@ -1319,6 +1364,56 @@ void dt_control_export(GList *imgid_list, int max_width, int max_height, int for
...
+  dt_control_image_enumerator_t *params
+      = (dt_control_image_enumerator_t *)calloc(1, sizeof(dt_control_image_enumerator_t));
+  if(!params)
+  {
+    dt_control_job_dispose(job);
+    return;
+  }
+  dt_control_job_set_params(job, params);
+  params->index = imgid_list;
+  dt_control_print_t *data = (dt_control_print_t*)malloc(sizeof(dt_control_print_t));
```
Comment (LebedevRI) [inline src/control/jobs/control_jobs.c]: if(!data)

File: src/libs/print_settings.c [hunk 14]
```diff
@@ -0,0 +1,1168 @@
...
+    {
+      g_strlcpy(printer_profile_filename, p->filename, sizeof(printer_profile_filename));
+      break;
+    }
+    prof = g_list_next(prof);
+  }
+
+  // the exported filename, on /tmp we do not want to mess with user's home directory
+
+  char filename[PATH_MAX];
```
Comment (LebedevRI) [inline src/libs/print_settings.c]: = {0};

File: src/libs/print_settings.c [hunk 15]
```diff
@@ -0,0 +1,1168 @@
...
+
+  snprintf(filename, PATH_MAX, "/tmp/pf_%d", imgid);
```
Comment (LebedevRI) [inline src/libs/print_settings.c]: sizeof()

File: src/libs/print_settings.c [hunk 16]
```diff
@@ -0,0 +1,1168 @@
...
+  else
+    gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON(ps->portrait), TRUE);
+}
+
+void
+gui_init (dt_lib_module_t *self)
+{
+  dt_lib_print_settings_t *d = (dt_lib_print_settings_t*)malloc(sizeof(dt_lib_print_settings_t));
+  self->data = d;
+  self->widget = gtk_table_new(18, 2, FALSE);
```
Comment (LebedevRI) [inline src/libs/print_settings.c]: GTK3: grid

File: src/libs/print_settings.c [hunk 17]
```diff
@@ -0,0 +1,1168 @@
...
+  g_signal_connect (G_OBJECT(d->lock_button), "toggled",
+                    G_CALLBACK(_lock_callback), self);
+
+  const gboolean lock_active = dt_conf_get_bool("plugins/print/print/lock_borders");
+  gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(d->lock_button), lock_active);
+
+  //// alignments
+
+  // Create the 3x3 gtk table toggle button table...
+  GtkTable *bat = GTK_TABLE(gtk_table_new(3,3,TRUE));
```
Comment (LebedevRI) [inline src/libs/print_settings.c]: GTK3: grid

File: src/libs/print_settings.c [hunk 18]
```diff
@@ -0,0 +1,1168 @@
...
+  d->profiles = g_list_append(d->profiles, prof);
+
+  prof = (dt_lib_export_profile_t *)g_malloc0(sizeof(dt_lib_export_profile_t));
+  g_strlcpy(prof->filename, "adobergb", sizeof(prof->filename));
+  dt_utf8_strlcpy(prof->name, _("Adobe RGB (compatible)"), sizeof(prof->name));
+  pos = prof->pos = 2;
+  d->profiles = g_list_append(d->profiles, prof);
+
+  // read datadir/color/out/*.icc
+  char datadir[PATH_MAX];
```
Comment (LebedevRI) [inline src/libs/print_settings.c]: = { 0 };

File: src/views/view.c [hunk 19]
```diff
@@ -69,7 +69,7 @@ void dt_view_manager_init(dt_view_manager_t *vm)
 #ifdef HAVE_MAP
                       "map",
 #endif
-                      "slideshow",  NULL };
+                      "slideshow",  "print", NULL };
```
Comment (LebedevRI) [inline src/views/view.c]: #ifdef HAVE_PRINT
         "print", 
#endif

File: CMakeLists.txt (hunk 4 above)
Comment (TurboGit) [inline CMakeLists.txt]: Sure, fixed.

File: src/common/cups_print.h (hunk 6 above)
Comment (TurboGit) [inline src/common/cups_print.h]: Right, cut&paste error. Fixed.

File: src/control/jobs/control_jobs.c (hunk 8 above)
Comment (TurboGit) [inline src/control/jobs/control_jobs.c]: Fixed.

File: src/control/jobs/control_jobs.c (hunk 9 above)
Comment (TurboGit) [inline src/control/jobs/control_jobs.c]: Fixed.

File: src/control/jobs/control_jobs.c (hunk 10 above)
Comment (TurboGit) [inline src/control/jobs/control_jobs.c]: Fixed too.

File: src/control/jobs/control_jobs.c (hunk 11 above)
Comment (TurboGit) [inline src/control/jobs/control_jobs.c]: Fixed.

File: src/control/jobs/control_jobs.c (hunk 12 above)
Comment (TurboGit) [inline src/control/jobs/control_jobs.c]: Fixed.

File: src/control/jobs/control_jobs.c (hunk 13 above)
Comment (TurboGit) [inline src/control/jobs/control_jobs.c]: Good catch! Fixed.

File: src/libs/print_settings.c (hunk 14 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Fixed.

File: src/libs/print_settings.c (hunk 15 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Fixed.

File: src/libs/print_settings.c (hunk 16 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: I'll let the expert convert to GTK3. houz has proposed to do that.

File: src/libs/print_settings.c (hunk 18 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Fixed.

File: src/views/view.c (hunk 19 above)
Comment (TurboGit) [inline src/views/view.c]: Indeed, fixed too.

File: src/common/printprof.c (hunk 7 above)
Comment (TurboGit) [inline src/common/printprof.c]: Right, I have used fprintf on stderr with "error:" as prefix as in many places already. will help dev to track down errors.

**Comment by TurboGit** [comment]
Thanks for all the comments. I have integrated all of them and pushed a new version.

File: src/libs/print_settings.c (hunk 5 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Now fixed also, I had forgotten it in the first batch of fixes.

**Comment by houz** [comment]
Could you merge in master so that I can see what I actually have to port to GTK3?

**Comment by TurboGit** [comment]
Now merged. Thanks for looking at this. I'll continue working on the print-module after.

**Comment by boucman** [comment]
ok, @hanatos meant to merge master into print-module, not the other way round

I have reverted the merge and rebased the print-module branch, hopefully I got it right.

feel free to reopen the PR and continue discussion there

**Comment by TurboGit** [comment]
Le dimanche 04 janvier 2015 à 08:23 -0800, Boucman a écrit :
> ok, @hanatos meant to merge master into print-module, not the other
> way round
> 
> I have reverted the merge and rebased the print-module branch,
> hopefully I got it right.
> 
> feel free to reopen the PR and continue discussion there

Ah! ok, sorry :( Thanks for reverting.

-- 
  Pascal Obry /  Magny Les Hameaux (78)

  The best way to travel is by means of imagination

  http://v2p.fr.eu.org
  http://www.obry.net

  gpg --keyserver keys.gnupg.net --recv-key F949BD3B

**Comment by TurboGit** [comment]
Note that to my defense the message said "merge in master" and I thought it was fine because one can always disable this module.

Anyway, I don't see a way to reopen this PR?

**Pull request merged**